### PR TITLE
Add interfaces to LF spec

### PR DESCRIPTION
**Pull request by sofiafaro-da**
Part of #11349. This PR adds the interface & implements declarations, deals with "requires", and name collisions. I'm leaving expression types and operational semantics to a future PR.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/spec/daml-lf-1.rst [hunk 1]
```diff
@@ -1588,6 +1607,23 @@ for the ``DefTemplate`` rule). ::
   ——————————————————————————————————————————————————————————————— DefException [Daml-LF ≥ 1.14]
     ⊢  'exception' T ↦ { 'message' e }
 
+    Mod:T  ∉  { Mod₁:T₁, …, Modₖ:Tₖ }
+    'interface' (x₁ : T₁) ↦ { 'requires' R₁ , … }  ∈ 〚Ξ〛Mod₁     R₁ ⊆ { Mod₁:T₁, …, Modₖ:Tₖ }
```
Comment (cocreature) [inline daml-lf/spec/daml-lf-1.rst:1611]: Can you help me understand your notation here? I would have expected that `requires` points to an interface so `R_1` is an interface. But then you say the interface is a subset of a set of templates?

Comment (cocreature) [reply daml-lf/spec/daml-lf-1.rst:1611]: Oh I see, `T_1` is not a template, it’s an interface and `R_1` is the set of all requires. That does make sense but I think it might be clearer if we use `I_1` for interfaces in place of templates.

**Review by cocreature — APPROVED** [review]
Nice! Apologies for the delay,t this somehow slipped through. Looks very sensible to me but I’d suggest to use different letters for interfaces & templates. Currently this is quite hard to follow at least for me.

File: daml-lf/spec/daml-lf-1.rst (hunk 1 above)
Comment (sofiafaro-da) [reply daml-lf/spec/daml-lf-1.rst:1611]: Yep, `R_1` is a set -- it seemed a lot easier to specify this way.
I agree with using `I` for interfaces, let me change that.

File: daml-lf/spec/daml-lf-1.rst [hunk 2]
```diff
@@ -778,8 +784,15 @@ available for usage::
             , 'agreement' e₄
             , 'choices' { ChDef₁, …, ChDefₘ }
             , KeyDef
+            , ImplDef₁, …, ImplDefₘ
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst:787]: we need another index than `ₘ`

File: daml-lf/spec/daml-lf-1.rst [hunk 3]
```diff
@@ -1541,15 +1558,15 @@ for the ``DefTemplate`` rule). ::
   Well-formed definitions │ ⊢  Def │
                           └────────┘
 
-     τ  ↠  τ₁'      αₙ : kₙ · … · α₁ : k₁  ⊢  τ₁'  :  ⋆
+     τ₁  ↠  τ₁'      αₙ : kₙ · … · α₁ : k₁  ⊢  τ₁'  :  ⋆
        ⋮
-     τ  ↠  τₘ'      αₙ : kₙ · … · α₁ : k₁  ⊢  τₘ'  :  ⋆
+     τₘ  ↠  τₘ'      αₙ : kₙ · … · α₁ : k₁  ⊢  τₘ'  :  ⋆
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst:1563]: good catch.

File: daml-lf/spec/daml-lf-1.rst [hunk 4]
```diff
@@ -1713,6 +1770,12 @@ name* construct as follows:
 * The *fully resolved name* of a choice ``Ch`` of a template
   definition ``'tpl' (x : T) ↦ { …, 'choices' { …, 'choice' ChKind Ch
   … ↦ …, … } }`` defined in the module ``Mod`` is ``Mod.T.Ch``.
+* The *fully resolved name* of an inherited choice ``Ch`` of a template
+  definition ``'tpl' (x : T) ↦ { …, 'implements' {  …, 'choices' { …, Ch,
+  … } } }`` defined in the module ``Mod`` is ``Mod.T.Ch``.
+* The *fully resolved name* of a choice ``Ch`` of an interface
+  definition ``'interface' (x : I) ↦ { …, 'choices' { …, 'choice' ChKind Ch
+  … ↦ …, … } }`` defined in the module ``Mod`` is ``Mod.I.Ch``.
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst:1778]: It maybe safer to include the method name.

**Review by remyhaemmerle-da — APPROVED** [review]
Nice start.

File: daml-lf/spec/daml-lf-1.rst (hunk 4 above)
Comment (sofiafaro-da) [reply daml-lf/spec/daml-lf-1.rst:1778]: That makes sense (e.g. don't want a collision between an interface choice and an interface method). I'll update the name collision checker separately.
